server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java

```diff
@@ -0,0 +1,291 @@
+
+            var middleOfStack = new ArrayList<>(testJvmStack);
+            middleOfStack.add(1, extraFrame);
+            assertIdleThreadHelper(idleThread, middleOfStack);
+        }
+    }
+
+    public void testSimilarity() {
+        var stackOne = makeThreadStackHelper(
+            List.of(
+                new String[]{"org.elasticsearch.monitor.test", "methodOne"},
+                new String[]{"org.elasticsearch.monitor.testOther", "methodTwo"}
+            )).toArray(new StackTraceElement[0]);
+
+        var stackTwo = makeThreadStackHelper(
+            List.of(
+                new String[]{"org.elasticsearch.monitor.test1", "methodOne"},
+                new String[]{"org.elasticsearch.monitor.testOther", "methodTwo"}
+            )).toArray(new StackTraceElement[0]);
+
+        var stackThree = makeThreadStackHelper(
+            List.of(
+                new String[]{"org.elasticsearch.monitor.testOther", "methodTwo"},
+                new String[]{"org.elasticsearch.monitor.test", "methodOne"}
+            )).toArray(new StackTraceElement[0]);
+
+        var stackFour = makeThreadStackHelper(
+            List.of(
+                new String[]{"org.elasticsearch.monitor.testPrior", "methodOther"},
+                new String[]{"org.elasticsearch.monitor.test", "methodOne"},
+                new String[]{"org.elasticsearch.monitor.testOther", "methodTwo"}
+            )).toArray(new StackTraceElement[0]);
+
+        HotThreads hotThreads = new HotThreads();
+
+        // We can simplify this with records when the toolchain is upgraded
+        class SimilarityTestCase {
+            final StackTraceElement[] one;
+            final StackTraceElement[] two;
+            final int similarityScore;
+
+            SimilarityTestCase(StackTraceElement[] one, StackTraceElement[] two, int similarityScore) {
+                this.one = one;
+                this.two = two;
+                this.similarityScore = similarityScore;
+            }
+        }
+
+        var testCases = new SimilarityTestCase[] {
+            new SimilarityTestCase(stackOne, stackOne, 2),
+            new SimilarityTestCase(stackOne, stackTwo, 1),
+            new SimilarityTestCase(stackOne, stackThree, 0),
+            new SimilarityTestCase(stackOne, stackFour, 2),
+            new SimilarityTestCase(stackTwo, stackFour, 1),
+            new SimilarityTestCase(stackOne, new StackTraceElement[0], 0),
+        };
+
+        for (var testCase : testCases) {
+            ThreadInfo threadOne = mock(ThreadInfo.class);
+            when(threadOne.getThreadName()).thenReturn("Thread One");
+            when(threadOne.getStackTrace()).thenReturn(testCase.one);
+
+            ThreadInfo threadTwo = mock(ThreadInfo.class);
+            when(threadTwo.getThreadName()).thenReturn("Thread Two");
+            when(threadTwo.getStackTrace()).thenReturn(testCase.two);
+
+            assertEquals(hotThreads.similarity(threadOne, threadTwo), testCase.similarityScore);
```

Comment:
`testCase.testCase.similarityScore` should be the expectation?

Reply:
Good catch! Thanks!